pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -2551,29 +2551,46 @@ public void readEntryFailed(ManagedLedgerException exception, Object ctx) {
                     }, null);
             return future;
         } else {
-            Long ledgerId = ((ManagedCursorContainer) ledger.getCursors()).getSlowestReaderPosition().getLedgerId();
+            PositionImpl slowestPosition = ((ManagedCursorContainer) ledger.getCursors()).getSlowestReaderPosition();
             try {
-                org.apache.bookkeeper.mledger.proto.MLDataFormats.ManagedLedgerInfo.LedgerInfo
-                        ledgerInfo = ledger.getLedgerInfo(ledgerId).get();
-                if (ledgerInfo != null && ledgerInfo.hasTimestamp() && ledgerInfo.getTimestamp() > 0
-                        && ((ManagedLedgerImpl) ledger).getClock().millis() - ledgerInfo.getTimestamp()
-                        > backlogQuotaLimitInSecond * 1000) {
-                    if (log.isDebugEnabled()) {
-                        log.debug("Time based backlog quota exceeded, quota {}, age of ledger "
-                                        + "slowest cursor currently on {}", backlogQuotaLimitInSecond * 1000,
-                                ((ManagedLedgerImpl) ledger).getClock().millis() - ledgerInfo.getTimestamp());
-                    }
-                    return CompletableFuture.completedFuture(true);
-                } else {
-                    return CompletableFuture.completedFuture(false);
-                }
+                return slowestReaderTimeBasedBacklogQuotaCheck(slowestPosition);
             } catch (Exception e) {
                 log.error("[{}][{}] Error reading entry for precise time based backlog check", topicName, e);
                 return CompletableFuture.completedFuture(false);
             }
         }
     }
 
+    private CompletableFuture<Boolean> slowestReaderTimeBasedBacklogQuotaCheck(PositionImpl slowestPosition)
+            throws ExecutionException, InterruptedException {
+        int backlogQuotaLimitInSecond = getBacklogQuota(BacklogQuotaType.message_age).getLimitTime();
+        Long ledgerId = slowestPosition.getLedgerId();
+        if (((ManagedLedgerImpl) ledger).getLedgersInfo().lastKey().equals(ledgerId)) {
+            return CompletableFuture.completedFuture(false);
+        }
+        int result;
+        org.apache.bookkeeper.mledger.proto.MLDataFormats.ManagedLedgerInfo.LedgerInfo
+                ledgerInfo = ledger.getLedgerInfo(ledgerId).get();
+        if (ledgerInfo != null && ledgerInfo.hasTimestamp() && ledgerInfo.getTimestamp() > 0
+                && ((ManagedLedgerImpl) ledger).getClock().millis() - ledgerInfo.getTimestamp()
+                > backlogQuotaLimitInSecond * 1000 && (result = slowestPosition.compareTo(
+                new PositionImpl(ledgerInfo.getLedgerId(), ledgerInfo.getEntries() - 1))) <= 0) {
+            if (result < 0) {
```

Comment:
Could you simplify the `if` clause? This `if` clause is like

```java
int result;
if (/* (1)... */ && (result = /* (2)... */) <= 0) {
    if (result < 0) {
        /* ... */
    } else { // NOTE: actually it means result == 0
        /* ... */
    }
} else {
    /* ... */
}
```

The readability is not good. IMO it's better to write

```java
if (/* (1)... */) {
    int result = /* (2)... */; // returned by `compareTo`
    if (result < 0) {
        /* ... */
    } else if (result == 0) {
        /* ... */
    } else {
        /* ... */
    }
}
```

Reply:
yes, I know, but it will add a `else` block.

Reply:
For removing duplicate code purpose, so I change like this.

Reply:
Just my suggestion, not a requested change.

From my personal perspective, I think the C-style assignment semantic harms the readability more than an extra `else` block.

```java
int result;
if ((result = doSomething()) <= 0) {
    // ...
}
```

is worse than

```java
int result = doSomething();
if (result <= 0) {
    // ...
}
```

In modern languages like Golang and Rust, the assignment `x = y` won't return anything, while Java adopts the same semantic from C, `x = y` returns the assigned `x`.

Reply:
> Just my suggestion, not a requested change.
> 
> From my personal perspective, I think the C-style assignment semantic harms the readability more than an extra `else` block.
> 
> ```java
> int result;
> if ((result = doSomething()) <= 0) {
>     // ...
> }
> ```
> 
> is worse than
> 
> ```java
> int result = doSomething();
> if (result <= 0) {
>     // ...
> }
> ```
> 
> In modern languages like Golang and Rust, the assignment `x = y` won't return anything, while Java adopts the same semantic from C, `x = y` returns the assigned `x`.

Only for my case,  if I add `else` block, the `else` block content will be redundant. It's not hard to read I think.

Reply:
Yes. I also found it. So I think it's controversial here and marked it as resolved.